### ISSUE TITLE
TypeError Issue with allennlp==0.8.4

### ISSUE THREAD

**Issue by yongzx** [BatsResearch/wiser#28]
It is not possible to import wiser `rules` and `dataset` which uses allennlp==0.8.4 because importing from `allennlp.data` returns the following error: 

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-060d3f614cdb> in <module>()
----> 1 from wiser.data.dataset_readers import NCBIDiseaseDatasetReader
      2 from wiser.rules import TaggingRule, LinkingRule, UMLSMatcher, DictionaryMatcher
      3 from wiser.generative import get_label_to_ix, get_rules
      4 from labelmodels import *
      5 from wiser.generative import train_generative_model

13 frames
/usr/local/lib/python3.7/dist-packages/wiser/data/__init__.py in <module>()
----> 1 from .util import *
      2 from .dataset_readers import *

/usr/local/lib/python3.7/dist-packages/wiser/data/util.py in <module>()
----> 1 from allennlp.data.fields import ArrayField
      2 from allennlp.data import Instance
      3 import numpy as np
      4 import pickle
      5 

/usr/local/lib/python3.7/dist-packages/allennlp/data/__init__.py in <module>()
----> 1 from allennlp.data.dataset_readers.dataset_reader import DatasetReader
      2 from allennlp.data.fields.field import DataArray, Field
      3 from allennlp.data.instance import Instance
      4 from allennlp.data.iterators.data_iterator import DataIterator
      5 from allennlp.data.token_indexers.token_indexer import TokenIndexer, TokenType

/usr/local/lib/python3.7/dist-packages/allennlp/data/dataset_readers/__init__.py in <module>()
      8 
      9 # pylint: disable=line-too-long
---> 10 from allennlp.data.dataset_readers.ccgbank import CcgBankDatasetReader
     11 from allennlp.data.dataset_readers.conll2003 import Conll2003DatasetReader
     12 from allennlp.data.dataset_readers.conll2000 import Conll2000DatasetReader

/usr/local/lib/python3.7/dist-packages/allennlp/data/dataset_readers/ccgbank.py in <module>()
      5 from overrides import overrides
      6 
----> 7 from allennlp.common.checks import ConfigurationError
      8 from allennlp.common.file_utils import cached_path
      9 from allennlp.data.dataset_readers.dataset_reader import DatasetReader

/usr/local/lib/python3.7/dist-packages/allennlp/common/__init__.py in <module>()
----> 1 from allennlp.common.from_params import FromParams
      2 from allennlp.common.params import Params
      3 from allennlp.common.registrable import Registrable
      4 from allennlp.common.tee_logger import TeeLogger
      5 from allennlp.common.tqdm import Tqdm

/usr/local/lib/python3.7/dist-packages/allennlp/common/from_params.py in <module>()
     46 
     47 from allennlp.common.checks import ConfigurationError
---> 48 from allennlp.common.params import Params
     49 
     50 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name

/usr/local/lib/python3.7/dist-packages/allennlp/common/params.py in <module>()
    171 
    172 
--> 173 class Params(MutableMapping):
    174     """
    175     Represents a parameter dictionary with a history, and contains other functionality around

/usr/local/lib/python3.7/dist-packages/allennlp/common/params.py in Params()
    234 
    235     @overrides
--> 236     def pop(self, key: str, default: Any = DEFAULT) -> Any:
    237         """
    238         Performs the functionality associated with dict.pop(key), along with checking for

/usr/local/lib/python3.7/dist-packages/overrides/overrides.py in overrides(method, check_signature, check_at_runtime)
     65     """
     66     if method:
---> 67         return _overrides(method, check_signature, check_at_runtime)
     68     else:
     69         return functools.partial(

/usr/local/lib/python3.7/dist-packages/overrides/overrides.py in _overrides(method, check_signature, check_at_runtime)
     91                 return wrapper  # type: ignore
     92             else:
---> 93                 _validate_method(method, super_class, check_signature)
     94                 return method
     95     raise TypeError(f"{method.__qualname__}: No super class method found")

/usr/local/lib/python3.7/dist-packages/overrides/overrides.py in _validate_method(method, super_class, check_signature)
    112         and not isinstance(super_method, property)
    113     ):
--> 114         ensure_signature_is_compatible(super_method, method, is_static)
    115 
    116 

/usr/local/lib/python3.7/dist-packages/overrides/signature.py in ensure_signature_is_compatible(super_callable, sub_callable, is_static)
     85         )
     86         ensure_all_positional_args_defined_in_sub(
---> 87             super_sig, sub_sig, super_type_hints, sub_type_hints, is_static, method_name
     88         )
     89     ensure_no_extra_args_in_sub(super_sig, sub_sig, is_static, method_name)

/usr/local/lib/python3.7/dist-packages/overrides/signature.py in ensure_all_positional_args_defined_in_sub(super_sig, sub_sig, super_type_hints, sub_type_hints, check_first_parameter, method_name)
    211         ):
    212             raise TypeError(
--> 213                 f"`{method_name}: {sub_param.name} must be a supertype of `{super_param.annotation}` but is `{sub_param.annotation}`"
    214             )
    215 

TypeError: `Params.pop: key must be a supertype of `<class 'inspect._empty'>` but is `<class 'str'>`
```

Environment: Python 3.7.10

**Comment by safranchik** [comment]
Hello!

Can you please try upgrading to AllenNLP=0.9.0, and replacing line 1 in `wiser/data/util.py` with `from allennlp.data.fields.array_field import ArrayField`?

**Comment by david-waterworth** [comment]
I'm seeing the same issue (but with the tutorial Intro Tutorial 1 - Tagging and Linking Rules.ipynb)

I upgraded to AllenNLP=0.9.0, the issue seems to be the `overrides` package performing validation, for example it complained that the `empty_field` property of the IndexField type in allennlp returns None. I fixed this by changing the type annotation

    @overrides
    def empty_field(self):
        return IndexField(-1, self.sequence_field.empty_field())

to

    @overrides
    def empty_field(self) -> 'IndexField':
        return IndexField(-1, self.sequence_field.empty_field())

I fixed this in a couple of places but ran into another

> TypeError: `SpacyTokenIndexer.tokens_to_indices: tokens must be a supertype of `typing.List[allennlp.data.tokenizers.token.Token]` but is `typing.List[spacy.tokens.token.Token]`

Again this appears to be another validation performed by `overrides`. I'll try and create a minimal example in a new virtualenv separate to wiser as it may be an allennlp issue?

**Comment by david-waterworth** [comment]
@safranchik I'd recommend you don't set a specific version for dependencies in requirements.txt, instead use

allennlp >= 0.8.4

If you want the ability to create a (mostly) reproduce-able environment I'd separately create a requirements-frozen.txt from

> pip freeze

That will pin not only the dependencies of wiser, but the dependencies of the dependencies. Pinning allennlp to 0.8.4 will quickly result in users ending up with untested dependency trees - i.e. an older version of allennlp with the latest version of it's dependencies.

**Comment by david-waterworth** [comment]
Note this was fixed in allennlp 1.0 (they pinned overrides), and the current release is 2.4.0 so your dependencies are way out of date. I'd recommend updating to the latest allennlp.

**Comment by yongzx** [comment]
@david-waterworth Thanks! Your suggestion of `overrides` works! I will close this issue for now.
